src/ucp/wireup/wireup.c

```diff
@@ -745,6 +866,77 @@ ucp_wireup_process_request(ucp_worker_h worker, ucp_ep_h ep,
     ucp_ep_set_failed_schedule(ep, UCP_NULL_LANE, status);
 }
 
+static UCS_F_NOINLINE void
+ucp_wireup_process_promote_request(ucp_worker_h worker, ucp_ep_h msg_ep,
+                                   const ucp_wireup_msg_t *msg,
+                                   const ucp_unpacked_address_t *remote_address)
+{
+    unsigned ep_init_flags            = UCP_EP_INIT_CREATE_AM_LANE;
+    ucp_ep_h ep                       = msg_ep;
+    ucs_usage_tracker_h usage_tracker = worker->usage_tracker.handle;
+    ucp_wireup_msg_extended_t *extended;
+    ucs_status_t status;
+
+    UCP_WIREUP_MSG_CHECK(msg, ep, UCP_WIREUP_MSG_PROMOTE);
+    extended = ucs_derived_of(msg, ucp_wireup_msg_extended_t);
+
+    if (ep == NULL) {
+        /* Avoid EP creation in case no promotion is expected */
+        if (!ucs_usage_tracker_is_promotable(usage_tracker, extended->score)) {
+            return;
+        }
+
+        status = ucp_wireup_get_ep_by_conn_sn(worker, remote_address, msg,
+                                              ep_init_flags, &ep);
+        if (status != UCS_OK) {
+            ucs_debug("ep %p: promotion request from 0x%.8lX denied because no"
+                      " local EP matches remote connection",
+                      ep, msg->src_ep_id);
+            return;
+        }
+    }
+
+    /* Prevent redundant promotion if request was sent */
+    if ((ep->flags & UCP_EP_FLAG_CONNECT_REQ_QUEUED) ||
```

Comment:
maybe move this check to `ucp_wireup_should_ignore_request` ?

Reply:
it's relevant only for this function